File: kb_python/data_structures/kb_stream.py
import json

class KB_Stream:
    """
    A class to handle the status data for the knowledge base.
    """
    def __init__(self, kb_search):
        self.kb_search = kb_search
        self.conn = self.kb_search.conn
        self.cursor = self.kb_search.cursor 
    
    def find_stream_id(self, node_name, properties, node_path):
        """
        Find the node id for a given node name, properties, node path, and data.
        """
        print(node_name, properties, node_path)
        result = self.find_node_ids(node_name, properties, node_path)
        if len(result) == 0:
            raise ValueError(f"No node found matching path parameters: {node_name}, {properties}, {node_path}")
        if len(result) > 1:
            raise ValueError(f"Multiple nodes found matching path parameters: {node_name}, {properties}, {node_path}")
        return result
    
    def find_stream_ids(self, node_name, properties, node_path):
        """
        Find the node id for a given node name, properties, node path :
        """
        print(node_name, properties, node_path)
        self.kb_search.clear_filters()
        self.kb_search.search_label("KB_STREAM_FIELD")
        if node_name is not None:
            self.kb_search.search_name(node_name)
        if properties is not None:
            for key in properties:
                self.kb_search.search_property_value(key, properties[key])
        if node_path is not None:
            self.kb_search.search_path(node_path)
        node_ids = self.kb_search.execute_query()
        
        if node_ids is None:
            raise ValueError(f"No node found matching path parameters: {node_name}, {properties}, {node_path}")
        if len(node_ids) == 0:
            raise ValueError(f"No node found matching path parameters: {node_name}, {properties}, {node_path}")
        return node_ids
    
    def find_stream_table_keys(self, key_data):
       
            
        return_values = []
        for key in key_data:
            
            return_values.append(key[5])
        return return_values
    

    
    def push_stream_data(self,
                         path: str,
                         data: dict,
                         max_retries: int = 3,
                         retry_delay: float = 1.0
                         ) -> int:
        """
        Find the earliest record (by recorded_at) for the given path,
        update it with new data and a fresh timestamp, and return its ID.

        Args:
            path (str): The path in LTREE format.
            data (dict): The JSON-serializable data to write.
            max_retries (int): Max attempts if rows are locked.
            retry_delay (float): Seconds to wait between retries.

        Returns:
            int: The ID of the updated record.

        Raises:
            Exception: If no records exist for this path.
            RuntimeError: If all matching rows are locked after retries.
        """
        for attempt in range(1, max_retries + 1):
            with self.conn.cursor() as cur:
                # 1) ensure there's at least one record to update
                cur.execute("""
                    SELECT COUNT(*)
                      FROM stream_table.stream_table
                     WHERE path = %s
                """, (path,))
                total = cur.fetchone()[0]
                if total == 0:
                    raise Exception(f"No records found for path={path!r}")

                # 2) try to lock the oldest candidate
                cur.execute("""
                    SELECT id
                      FROM stream_table.stream_table
                     WHERE path = %s
                     ORDER BY recorded_at ASC
                     FOR UPDATE SKIP LOCKED
                     LIMIT 1
                """, (path,))
                row = cur.fetchone()

                if not row:
                    # rows exist but all are currently locked
                    self.conn.rollback()
                    if attempt < max_retries:
                        time.sleep(retry_delay)
                        continue
                    else:
                        raise RuntimeError(
                            f"Could not lock any row for path={path!r} after {max_retries} attempts"
                        )

                record_id = row[0]

                # 3) perform the update
                cur.execute("""
                    UPDATE stream_table.stream_table
                       SET data         = %s,
                           recorded_at  = NOW()
                     WHERE id = %s
                """, (json.dumps(data), record_id))

                if cur.rowcount != 1:
                    self.conn.rollback()
                    raise Exception(f"Failed to update record id={record_id}")

                self.conn.commit()
                return record_id

        # Should never reach here
        raise RuntimeError("Unexpected error in push_stream_data")
    

        
    def list_stream_data(self, path, limit=None, offset=0, recorded_after=None, recorded_before=None):
        """
        List all jobs for a given path where valid is False and is_active is False,
        ordered by completed_at with earliest first.
        
        Args:
            path (str): The path to search for in LTREE format
            limit (int, optional): Maximum number of jobs to return
            offset (int, optional): Number of jobs to skip
            completed_after (datetime, optional): Only include jobs completed after this time
            completed_before (datetime, optional): Only include jobs completed before this time
        
        Returns:
            list: A list of dictionaries containing all job details
            
        Raises:
            Exception: If there's an error executing the query
        """
        try:
            # Build the base query
            query = """
                SELECT id, path, recorded_at, data
                FROM stream_table.stream_table
                WHERE path = %s
                
            """
            
            params = [path]
            
            # Add optional time-based filters
            if recorded_after is not None:
                query += " AND recorded_at >= %s"
                params.append(recorded_after)
                
            if recorded_before is not None:
                query += " AND recorded_at <= %s"
                params.append(recorded_before)
                
            # Add ordering
            query += " ORDER BY recorded_at ASC"
            
            # Add optional pagination
            if limit is not None:
                query += " LIMIT %s"
                params.append(limit)
                
            if offset > 0:
                query += " OFFSET %s"
                params.append(offset)
                
            self.cursor.execute(query, params)
            results = self.cursor.fetchall()
            
            # Get column names from cursor description
            column_names = [desc[0] for desc in self.cursor.description]
            
            # Convert results to a list of dictionaries using column names
            jobs = []
            for row in results:
                job = dict(zip(column_names, row))
                jobs.append(job)
            
            return jobs
        
        except Exception as e:
            # This is a read-only operation, so no need to rollback
            raise Exception(f"Error stream data: {str(e)}")
        

File: kb_modules/kb_dotnet/postgres/data_structures/KnowledgeBase/KBStream.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Npgsql;
using Newtonsoft.Json;

public class StreamRecord
{
    public int Id { get; set; }
    public string Path { get; set; } = string.Empty;
    public DateTime RecordedAt { get; set; }
    public object Data { get; set; } = new();
    public bool Valid { get; set; }
}

public class ClearStreamDataResult
{
    public bool Success { get; set; }
    public int ClearedCount { get; set; }
    public List<dynamic>? ClearedRecords { get; set; }
    public string? Error { get; set; }
}

public class KBStream
{
    private readonly KBSearch _kbSearch;
    private readonly string _baseTable;

    public KBStream(KBSearch kbSearch, string database)
    {
        _kbSearch = kbSearch;
        _baseTable = $"{database}_stream";
    }

    private NpgsqlConnection GetConnection()
    {
        return _kbSearch.GetConnection();
    }

    private async Task<List<dynamic>> ExecuteQueryAsync(string query, List<object>? parameters = null)
    {
        parameters ??= new List<object>();
        var client = GetConnection();
        
        using var cmd = new NpgsqlCommand(query, client);
        for (int i = 0; i < parameters.Count; i++)
        {
            cmd.Parameters.AddWithValue(parameters[i]);
        }

        using var reader = await cmd.ExecuteReaderAsync();
        var results = new List<dynamic>();
        
        while (await reader.ReadAsync())
        {
            var row = new Dictionary<string, object>();
            for (int i = 0; i < reader.FieldCount; i++)
            {
                row[reader.GetName(i)] = reader.IsDBNull(i) ? DBNull.Value : reader.GetValue(i);
            }
            results.Add(row);
        }
        
        return results;
    }

    private async Task<dynamic?> ExecuteSingleAsync(string query, List<object>? parameters = null)
    {
        var results = await ExecuteQueryAsync(query, parameters);
        return results.FirstOrDefault();
    }

    private async Task SleepAsync(int milliseconds)
    {
        await Task.Delay(milliseconds);
    }

    public async Task<dynamic> FindStreamIdAsync(
        string? kb = null,
        string? nodeName = null,
        Dictionary<string, object>? properties = null,
        string? nodePath = null)
    {
        var results = await FindStreamIdsAsync(kb, nodeName, properties, nodePath);
        
        if (results.Count == 0)
        {
            throw new InvalidOperationException(
                $"No stream node found matching parameters: name={nodeName}, properties={JsonConvert.SerializeObject(properties)}, path={nodePath}");
        }
        
        if (results.Count > 1)
        {
            throw new InvalidOperationException(
                $"Multiple stream nodes ({results.Count}) found matching parameters: name={nodeName}, properties={JsonConvert.SerializeObject(properties)}, path={nodePath}");
        }
        
        return results[0];
    }

    public async Task<List<dynamic>> FindStreamIdsAsync(
        string? kb = null,
        string? nodeName = null,
        Dictionary<string, object>? properties = null,
        string? nodePath = null)
    {
        try
        {
            _kbSearch.ClearFilters();
            _kbSearch.SearchLabel("KB_STREAM_FIELD");
            
            if (!string.IsNullOrEmpty(kb)) _kbSearch.SearchKb(kb);
            if (!string.IsNullOrEmpty(nodeName)) _kbSearch.SearchName(nodeName);
            
            if (properties != null)
            {
                foreach (var kvp in properties)
                {
                    _kbSearch.SearchPropertyValue(kvp.Key, kvp.Value);
                }
            }
            
            if (!string.IsNullOrEmpty(nodePath)) _kbSearch.SearchPath(nodePath);

            var nodeIds = await _kbSearch.ExecuteQueryAsync();
            
            if (nodeIds == null || nodeIds.Count == 0)
            {
                throw new InvalidOperationException(
                    $"No stream nodes found matching parameters: name={nodeName}, properties={JsonConvert.SerializeObject(properties)}, path={nodePath}");
            }
            
            return nodeIds;
        }
        catch (Exception e)
        {
            if (e.Message.StartsWith("No stream")) throw;
            throw new Exception($"Error finding stream node IDs: {e.Message}", e);
        }
    }

    public List<string> FindStreamTableKeys(List<dynamic> rows)
    {
        if (rows == null || rows.Count == 0) return new List<string>();
        
        return rows
            .Select(r => ((Dictionary<string, object>)r)["path"]?.ToString())
            .Where(p => !string.IsNullOrEmpty(p))
            .Cast<string>()
            .ToList();
    }

    public async Task<Dictionary<string, object>> PushStreamDataAsync(
        string path,
        Dictionary<string, object> data,
        int maxRetries = 3,
        int retryDelay = 1000)
    {
        if (string.IsNullOrEmpty(path)) 
            throw new ArgumentException("Path cannot be empty");
        
        if (data == null) 
            throw new ArgumentException("Data must be an object");

        var client = GetConnection();

        for (int attempt = 1; attempt <= maxRetries; attempt++)
        {
            try
            {
                var countRow = await ExecuteSingleAsync(
                    $"SELECT COUNT(*) AS count FROM {_baseTable} WHERE path = $1::ltree",
                    new List<object> { path });
                
                var total = Convert.ToInt32(((Dictionary<string, object>)countRow)["count"]);
                
                if (total == 0)
                {
                    throw new InvalidOperationException($"No records found for path='{path}'. Must pre-allocate.");
                }

                using var transaction = await client.BeginTransactionAsync();
                
                try
                {
                    var row = await ExecuteSingleAsync(
                        $@"
                        SELECT id, recorded_at, valid
                        FROM {_baseTable}
                        WHERE path = $1::ltree
                        ORDER BY recorded_at ASC
                        FOR UPDATE SKIP LOCKED
                        LIMIT 1
                        ",
                        new List<object> { path });

                    if (row == null)
                    {
                        await transaction.RollbackAsync();
                        if (attempt < maxRetries)
                        {
                            await SleepAsync(retryDelay);
                            continue;
                        }
                        else
                        {
                            throw new InvalidOperationException(
                                $"Could not lock any row for path='{path}' after {maxRetries} attempts");
                        }
                    }

                    var rowDict = (Dictionary<string, object>)row;
                    var prevRecordedAt = (DateTime)rowDict["recorded_at"];
                    var wasValid = (bool)rowDict["valid"];
                    var id = Convert.ToInt32(rowDict["id"]);

                    var upd = await ExecuteSingleAsync(
                        $@"
                        UPDATE {_baseTable}
                        SET data = $1::json, recorded_at = NOW(), valid = TRUE
                        WHERE id = $2
                        RETURNING id, path, recorded_at, data, valid
                        ",
                        new List<object> { JsonConvert.SerializeObject(data), id });

                    if (upd == null)
                    {
                        await transaction.RollbackAsync();
                        throw new InvalidOperationException($"Failed to update record id={id}");
                    }

                    await transaction.CommitAsync();

                    var updDict = (Dictionary<string, object>)upd;
                    return new Dictionary<string, object>
                    {
                        ["id"] = updDict["id"],
                        ["path"] = updDict["path"],
                        ["recorded_at"] = updDict["recorded_at"],
                        ["data"] = updDict["data"],
                        ["valid"] = updDict["valid"],
                        ["previous_recorded_at"] = prevRecordedAt,
                        ["was_previously_valid"] = wasValid,
                        ["operation"] = "circular_buffer_replace"
                    };
                }
                catch
                {
                    await transaction.RollbackAsync();
                    throw;
                }
            }
            catch (Exception e)
            {
                if (e.Message.StartsWith("No records") || e.Message.Contains("Could not lock"))
                {
                    throw;
                }
                
                if (attempt < maxRetries)
                {
                    await SleepAsync(retryDelay);
                    continue;
                }
                else
                {
                    throw new Exception($"Error pushing stream data: {e.Message}", e);
                }
            }
        }
        
        throw new Exception("Unexpected error in pushStreamData");
    }

    public async Task<Dictionary<string, object>?> GetLatestStreamDataAsync(string path)
    {
        if (string.IsNullOrEmpty(path)) 
            throw new ArgumentException("Path cannot be empty");

        var row = await ExecuteSingleAsync(
            $@"
            SELECT id, path, recorded_at, data, valid
            FROM {_baseTable}
            WHERE path = $1::ltree AND valid = TRUE
            ORDER BY recorded_at DESC
            LIMIT 1
            ",
            new List<object> { path });

        return row as Dictionary<string, object>;
    }

    public async Task<int> GetStreamDataCountAsync(string path, bool includeInvalid = false)
    {
        if (string.IsNullOrEmpty(path)) 
            throw new ArgumentException("Path cannot be empty");

        var query = includeInvalid
            ? $"SELECT COUNT(*) AS count FROM {_baseTable} WHERE path = $1::ltree"
            : $"SELECT COUNT(*) AS count FROM {_baseTable} WHERE path = $1::ltree AND valid = TRUE";

        var row = await ExecuteSingleAsync(query, new List<object> { path });
        var rowDict = (Dictionary<string, object>)row;
        return Convert.ToInt32(rowDict["count"]);
    }

    public async Task<ClearStreamDataResult> ClearStreamDataAsync(string path, DateTime? olderThan = null)
    {
        if (string.IsNullOrEmpty(path)) 
            throw new ArgumentException("Path cannot be empty");

        try
        {
            string query;
            List<object> parameters;

            if (olderThan.HasValue)
            {
                query = $@"
                    UPDATE {_baseTable}
                    SET valid = FALSE
                    WHERE path = $1::ltree AND recorded_at < $2 AND valid = TRUE
                    RETURNING id, recorded_at
                ";
                parameters = new List<object> { path, olderThan.Value };
            }
            else
            {
                query = $@"
                    UPDATE {_baseTable}
                    SET valid = FALSE
                    WHERE path = $1::ltree AND valid = TRUE
                    RETURNING id, recorded_at
                ";
                parameters = new List<object> { path };
            }

            var records = await ExecuteQueryAsync(query, parameters);
            
            return new ClearStreamDataResult
            {
                Success = true,
                ClearedCount = records.Count,
                ClearedRecords = records
            };
        }
        catch (Exception e)
        {
            return new ClearStreamDataResult
            {
                Success = false,
                ClearedCount = 0,
                Error = e.Message
            };
        }
    }

    public async Task<List<dynamic>> ListStreamDataAsync(
        string path,
        int? limit = null,
        int offset = 0,
        DateTime? recordedAfter = null,
        DateTime? recordedBefore = null,
        string order = "ASC")
    {
        if (string.IsNullOrEmpty(path)) 
            throw new ArgumentException("Path cannot be empty");
        
        if (!new[] { "ASC", "DESC" }.Contains(order))
            throw new ArgumentException("Order must be 'ASC' or 'DESC'");

        var query = $@"
            SELECT id, path, recorded_at, data, valid
            FROM {_baseTable}
            WHERE path = $1::ltree AND valid = TRUE
        ";
        
        var parameters = new List<object> { path };

        if (recordedAfter.HasValue)
        {
            parameters.Add(recordedAfter.Value);
            query += $" AND recorded_at >= ${parameters.Count}";
        }

        if (recordedBefore.HasValue)
        {
            parameters.Add(recordedBefore.Value);
            query += $" AND recorded_at <= ${parameters.Count}";
        }

        query += $" ORDER BY recorded_at {order}";

        if (limit.HasValue && limit.Value > 0)
        {
            parameters.Add(limit.Value);
            query += $" LIMIT ${parameters.Count}";
        }

        if (offset > 0)
        {
            parameters.Add(offset);
            query += $" OFFSET ${parameters.Count}";
        }

        return await ExecuteQueryAsync(query, parameters);
    }

    public async Task<List<dynamic>> GetStreamDataRangeAsync(
        string path,
        DateTime startTime,
        DateTime endTime)
    {
        if (string.IsNullOrEmpty(path)) 
            throw new ArgumentException("Path cannot be empty");
        
        if (startTime >= endTime) 
            throw new ArgumentException("startTime must be before endTime");

        var query = $@"
            SELECT id, path, recorded_at, data, valid
            FROM {_baseTable}
            WHERE path = $1::ltree
              AND recorded_at >= $2
              AND recorded_at <= $3
              AND valid = TRUE
            ORDER BY recorded_at ASC
        ";

        return await ExecuteQueryAsync(query, new List<object> { path, startTime, endTime });
    }

    public async Task<Dictionary<string, object>> GetStreamStatisticsAsync(
        string path,
        bool includeInvalid = false)
    {
        if (string.IsNullOrEmpty(path)) 
            throw new ArgumentException("Path cannot be empty");

        var statsQuery = includeInvalid
            ? $@"
                SELECT 
                  COUNT(*) AS total_records,
                  COUNT(CASE WHEN valid THEN 1 END) AS valid_records,
                  COUNT(CASE WHEN NOT valid THEN 1 END) AS invalid_records,
                  MIN(CASE WHEN valid THEN recorded_at END) AS earliest_valid_recorded,
                  MAX(CASE WHEN valid THEN recorded_at END) AS latest_valid_recorded,
                  MIN(recorded_at) AS earliest_recorded_overall,
                  MAX(recorded_at) AS latest_recorded_overall,
                  AVG(EXTRACT(EPOCH FROM (recorded_at - LAG(recorded_at) OVER (ORDER BY recorded_at)))) AS avg_interval_seconds_all,
                  AVG(CASE WHEN valid THEN EXTRACT(EPOCH FROM (recorded_at - LAG(recorded_at) OVER (ORDER BY recorded_at))) END) AS avg_interval_seconds_valid
                FROM {_baseTable}
                WHERE path = $1::ltree
              "
            : $@"
                SELECT 
                  COUNT(*) AS valid_records,
                  MIN(recorded_at) AS earliest_recorded,
                  MAX(recorded_at) AS latest_recorded,
                  AVG(EXTRACT(EPOCH FROM (recorded_at - LAG(recorded_at) OVER (ORDER BY recorded_at)))) AS avg_interval_seconds
                FROM {_baseTable}
                WHERE path = $1::ltree AND valid = TRUE
              ";

        var row = await ExecuteSingleAsync(statsQuery, new List<object> { path });
        return (row as Dictionary<string, object>) ?? new Dictionary<string, object>();
    }

    public async Task<Dictionary<string, object>?> GetStreamDataByIdAsync(int recordId)
    {
        if (!int.TryParse(recordId.ToString(), out _))
        {
            throw new ArgumentException("recordId must be an integer");
        }

        var query = $@"
            SELECT id, path, recorded_at, data
            FROM {_baseTable}
            WHERE id = $1
        ";

        var result = await ExecuteSingleAsync(query, new List<object> { recordId });
        return result as Dictionary<string, object>;
    }

    public async Task DisconnectAsync()
    {
        // Connection is managed by KBSearch, so this is a no-op
        await Task.CompletedTask;
    }
}